rfcs/0019-storing-client-data.md

```diff
@@ -0,0 +1,541 @@
+    amount: Unsigned64.t;
+    fee: Unsigned64.t;
+    memo: string;
+    receipt_chain_hash: string
+  }
+
+end
+
+module Fee_transfer {
+  id: string;
+  fee: Int64.t;
+  receiver: Public_key.Compressed.t; 
+}
+```
+
+You may notice that the user_commands type has the field receipt_chain_hash. This is a reference to a `receipt_chain_hash` which is used to prove the payment of a transaction. The API for proving the existence of a payment, `prove_payment`, would now require `payment_id` as an argument (See Appendix for the API of `prove_payment`). There is also an `addPaymentReceipt` API that allows a client to add an arbitrary transaction along with the date that the client saw the transaction. However, the transaction input itself cannot just be used to prove that a transaction has been sent. To prove a transaction, we need use current receipt_chain_hash of a transaction and its previous receipt_chain_hash. Therefore, inputs for the `addPaymentReceipt` API should be a receipt_chain and its parent.
+
+Below is the type of a `receipt_chain_hash`:
+
+```ocaml
+module Receipt_chain_hash = struct
+  type t = {
+    hash: string
+    previous_hash: string
+  }
+end
+```
+
+Additionally, clients should be able to know which external_transitions contains a transaction. These blocks can gives us useful information of a transaction, like the number of block confirmations that is on top of a certain transaction and the "status" of a transaction. Therefore, a client should be able to store a record of these interested external_transitions and look them up easily in a container. Here is the type of an external_transition:
+
+```ocaml
+module Protocol_state = struct
+  type t = {
+    previousStateHash: string;
+    blockchainState: BlockchainState;
+    length: Int64!
+  }
+end
+
+
+module External_transition = struct
+  state_hash: string;
+  creator: Public_key.Compressed.t
+  protocol_state: Protocol_state.t [not null]
+  is_snarked bool [not null]
+  status consensus_status [not null]
+  time time [not null]
+end
+```
+
+The OCaml implementation of `External_transition` is different from the presented `external_transition` type as it has the field `is_snarked` and `status` to make it easy to compute the `transactionStatus` query, which will tell us if a transaction is snarked and it's `consensus_status`. `consensus_status` is variant type and is described below:
+
+```ocaml
+type consensus_status =
+  | Failed (* frontier removed transition and never reached finality *)
+  | Confirmed (* Transition reached finality by passing the root *)
+  | Pending of int (* Transition is somewhere in the frontier and has a block confirmation number *)
+  | Unknown (* Could not compute status of transaction i.e. We don't have a record of it *)
+```
+
+Notice that `consensus_status` is a block's position in Nakamoto consensus. This position has nothing to do with a block being snarked or not, so `Snarked` is not a constructor of this variant. 
```

Comment:
Nakamoto consensus? What?

Reply:
Lol. Nakamoto consensus is basically having your consensus state represented as tree, like in Bitcoin and our protocol. Other distributed systems consensus algorithms use some flavor of Byzantine consensus, which is a pointer of the current state of the system and a pointer to the next candidate state. 

So in the context of this RFC, Nakamoto consensus represents the node in the consensus tree.